vip-jetpack/connection-pilot/class-jetpack-connection-status-check.php

```diff
@@ -0,0 +1,199 @@
+	 */
+	private function handle_disconnects_using_last_healthcheck() {
+		if ( $this->healthcheck_option['site_url'] === $this->site_url ) {
+			$reconnection_attempt = $this->connect_site();
+			return $this->notify_pilot( $reconnection_attempt );
+		}
+
+		return $this->notify_pilot( 'Jetpack is disconnected, and it appears the domain has changed.' );
+	}
+
+	/**
+	 * Connection issue handler for new sites.
+	 *
+	 * This includes either a placeholder go-vip domain or a recently created multi-site site.
+	 */
+	private function handle_new_sites() {
+		if ( $this->is_placeholder_domain() ) {
+			return $this->notify_pilot( 'Jetpack is disconnected, though it appears this is a new site.' );
+		}
+
+		// Must be a new site on a MS.
+		return $this->notify_pilot( 'Jetpack is disconnected, though it appears this is a new site on a MS network.' );
+	}
+
+	/**
+	 * Try to (re)connect the site.
+	 *
+	 * @return string A message with the result of the (re)connection attempt.
+	 */
+	private function connect_site() {
+		// Skip the JP connection tests since we've already run them.
+		$connection_attempt = WPCOM_VIP_Jetpack_Connection_Controls::connect_site( 'skip_connection_tests' );
+
+		if ( true === $connection_attempt ) {
+			if ( ! empty( $this->healthcheck_option['cache_site_id'] ) && (int) Jetpack_Options::get_option( 'id' ) !== (int) $this->healthcheck_option['cache_site_id'] ) {
+				return 'Alert: Jetpack was automatically reconnected, but the connection may have changed cache sites. Needs manual inspection.';
+			}
+			return 'Jetpack was successfully (re)connected!';
+		}
+
+		if ( is_wp_error( $connection_attempt ) ) {
+			$this->connection_error = $connection_attempt;
+		}
+
+		return 'Jetpack (re)connection attempt failed.';
+	}
+
+	/**
+	 * Check if the current site is using a placeholder VIP Go domain name.
+	 *
+	 * @return bool True if this is a placeholder URL.
+	 */
+	private function is_placeholder_domain() {
+		$site_parsed = wp_parse_url( $this->site_url );
+		return wp_endswith( $site_parsed['host'], '.go-vip.co' ) || wp_endswith( $site_parsed['host'], '.go-vip.net' );
+	}
+
+	/**
+	 * Check if the current site is on a MS install and was recently added.
+	 *
+	 * @return bool True if this is a newly created site.
+	 */
+	private function is_new_multisite_site() {
+		if ( ! is_multisite() ) {
+			return false;
+		}
+
+		$current_site = get_blog_details();
+		if ( empty( $current_site->registered ) ) {
+			return false;
+		}
+
+		$time_diff = time() - strtotime( $current_site->registered );
+		if ( 2 >= ( $time_diff / 3600 ) ) {
+			// The site was created in the last 2 hours. The cron job will run hourly, so this should catch most cases.
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Adds to the array of notifications that need to be sent out by the Pilot.
+	 *
+	 * @param string $message The first line of the notification that will be sent.
+	 */
+	private function notify_pilot( $message ) {
```

Comment:
IMO it would be better to pass the error object into `notify_pilot()` explicitly, rather than using the class property. Using the single property gets messy because something else could overwrite it at any time, at which point the message and underlying error can be different things.

Better to be explicit and pass the error in as an argument at runtime.